lib/Crypto/doc/sailfish-crypto-overview.qdoc

```diff
@@ -0,0 +1,218 @@
+} else {
+    symmetricKeyReference = generateReply.argumentAt<1>();
+}
+\endcode
+
+\section4 Encrypting data with a symmetric key
+
+After generating a symmetric key, that key may be used to encrypt data.
+Note that the key may be either a key reference (that is, a key which contains
+only metadata and a valid identifier, which references a full key stored in
+secure storage) or a full key (that is, a key which contains secret key data).
+
+In this example, we use the reference key which was returned to the application
+in the previous snippet.
+
+\code
+QByteArray ciphertext, plaintext = "Example plaintext data";
+QDBusPendingReply<Sailfish::Crypto::Result, QByteArray>
+        encryptReply = cm.encrypt(
+                plaintext,
+                symmetricKeyReference,
+                Sailfish::Crypto::Key::BlockModeCBC,
+                Sailfish::Crypto::Key::EncryptionPaddingNone,
+                Sailfish::Crypto::Key::DigestSha256,
+                Sailfish::Crypto::CryptoManager::DefaultCryptoStoragePluginName);
+encryptReply.waitForFinished();
+if (encryptReply.argumentAt<0>().code() == Sailfish::Crypto::Result::Failed) {
+    qWarning() << "Failed to encrypt:"
+               << encryptReply.argumentAt<0>().errorMessage();
+} else {
+    ciphertext = encryptReply.argumentAt<1>();
+}
+\endcode
+
+\section4 Decrypting data with a symmetric key
+
+A symmetric key may also be used to decrypt data.  In the following snippet,
+the client asks the system service to decrypt the ciphertext with the same
+key reference, to produce decrypted data.
+
+\code
+QDBusPendingReply<Sailfish::Crypto::Result, QByteArray>
+        decryptReply = cm.decrypt(
+                ciphertext,
+                symmetricKeyReference,
+                Sailfish::Crypto::Key::BlockModeCBC,
+                Sailfish::Crypto::Key::EncryptionPaddingNone,
+                Sailfish::Crypto::Key::DigestSha256,
+                Sailfish::Crypto::CryptoManager::DefaultCryptoStoragePluginName);
+decryptReply.waitForFinished();
+if (decryptReply.argumentAt<0>().code() == Sailfish::Crypto::Result::Failed) {
+    qWarning() << "Failed to decrypt ciphertext:"
+               << decryptReply.argumentAt<0>().errorMessage();
+} else {
+    QByteArray decrypted = decryptReply.argumentAt<1>();
+    qDebug() << "Decrypted:" << decrypted;
+}
+\endcode
+
+\section1 Extending the Sailfish OS Secrets and Crypto Framework with Crypto Plugins
+
+The Sailfish OS Crypto Library also provides a plugin base-class which may be
+extended by device vendors or trusted partners to allow them to build plugins
+to extend the Sailfish OS Secrets and Crypto Framework with additional
+cryptography functionality (for example, supporting different algorithms or
+block modes, or performing the operations via a Trusted Execution Environment
+application rather than in-process in the rich application process).
+
+The \l{Sailfish::Crypto::CryptoPlugin} class should be extended in order to
```

Comment:
Did we decide how we handle custom algorithms, which are not listed in the enumeration?

Reply:
Not yet, it is an open question.  In my opinion, we could add another enum PluginCustomAlgorithm (perhaps with value 1, i.e. next after UnknownAlgorithm), and document it to say that if that algorithm is specified, a plugin-specific CustomParameter must be provided to tell the plugin which algorithm to use.  Does anyone object to that idea, or have any other suggestions?